### PR TITLE
DWR-443 - part 1

### PR DESCRIPTION
**Pull request by agorina**
What is included from the JIRA:
Revisit all builders in the exam-processor to check for valid conditions, including but not limited to:
- [x] null checks
- [ ] data size, e.g. VARCHAR field limits
- [x] grade (this is meant for the asmt to test verification)
- [x] subject (same as above)
- [x] scale score min/max range
- [x] add check for non-null sessionId
- [x] Establish the pattern of validation and make it consistent.
- [ ] These changes may apply to package processor as well. Either make those changes as part of this, or create a new issue if scope is too large.

### CODE REVIEW

File: common/src/main/java/org/opentestsystem/rdw/ingest/common/util/ParserHelper.java [hunk 1]
```diff
@@ -10,63 +13,117 @@
     private final DataElementErrorCollector elementErrorCollector;
 
     /**
-     * This is an optional construction for use only when you don't care about collecting errors
+     * General construction that collects errors in to a {@link DataElementErrorCollector}
+     *
+     * @param elementErrorCollector error collector to use for any errors
      */
-    public ParserHelper() {
-        this.elementErrorCollector = null;
```
Comment (agorina) [inline common/src/main/java/org/opentestsystem/rdw/ingest/common/util/ParserHelper.java]: This was not used.

File: common/src/main/java/org/opentestsystem/rdw/ingest/common/util/Precondition.java [hunk 2]
```diff
@@ -18,4 +18,7 @@ public static String checkNotBlank(final String value) {
         checkArgument(!isNullOrEmpty(value), "value may not be blank");
         return value;
     }
+
+    private Precondition() {
+    }
```
Comment (agorina) [inline common/src/main/java/org/opentestsystem/rdw/ingest/common/util/Precondition.java]: Unrelated, just made it consistent with other utility classes.

File: exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Assessment.java [hunk 3]
```diff
@@ -20,8 +20,11 @@
     private String name;
     private String label;
     private int gradeId;
+    private String gradeCode;
     private int typeId;
+    private String typeCode;
     private int subjectId;
+    private String subjectCode;
```
Comment (agorina) [inline exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Assessment.java]: I need it to compare to the Test section of the exam.

File: exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Exam.java [hunk 4]
```diff
@@ -20,7 +19,7 @@
     private Instant completedAt;
     private long assessmentId;
     private String asmtVersion;
-    private Long schoolYear;
+    private int schoolYear;
```
Comment (agorina) [inline exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Exam.java]: This should have been int to begin with.

File: exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Exam.java [hunk 5]
```diff
@@ -139,9 +138,6 @@ public static Builder builder() {
         private StudentExamAttributes studentExamAttributes;
 
         public Exam build() {
-            checkArgument(administrationConditionId != null, "administrationConditionId may not be null");
-            checkArgument(completenessId != null, "completenessId may not be null");
-
             final Exam exam = new Exam();
```
Comment (agorina) [inline exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Exam.java]: Builders are simple data collectors. All validation is done in the services.

File: exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/StudentGroup.java [hunk 6]
```diff
@@ -39,10 +38,6 @@ public int getSchoolYear() {
         return schoolYear;
     }
 
-    public int getSubjectId() {
-        return subjectId;
-    }
-
```
Comment (agorina) [inline exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/StudentGroup.java]: This is part of DWR-511, we are not going to inferre the group subject from the exam.

File: exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcAssessmentRepository.java [hunk 7]
```diff
@@ -38,7 +39,6 @@
 
     @Transactional(readOnly = true)
     @Override
-    //TODO: consider caching here
```
Comment (agorina) [inline exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcAssessmentRepository.java]: I entered a JIRA for this.

File: exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultAssessmentService.java [hunk 8]
```diff
@@ -14,13 +18,28 @@
...
+
+        final DataElementErrorCollector errorCollector = new DataElementErrorCollector();
+        if (!assessment.getSubjectCode().equalsIgnoreCase(test.getSubject()))
+            errorCollector.add(new DataElementError("Assessment natural id [" + assessmentNaturalId + "], Test subject", test.getSubject(), "exam subject does not match asmt subject"));
+
+        if (!assessment.getTypeCode().equalsIgnoreCase(test.getAssessmentType()))
+            errorCollector.add(new DataElementError("Assessment natural id [" + assessmentNaturalId + "], Test assessmentType", test.getAssessmentType(), "exam subject does not match asmt type"));
+
+        if (!assessment.getGradeCode().equalsIgnoreCase(test.getGrade()))
+            errorCollector.add(new DataElementError("Assessment natural id [" + assessmentNaturalId + "], Test grade", test.getGrade(), "exam subject does not match asmt grade"));
```
Comment (agorina) [inline exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultAssessmentService.java]: I believe this is verification is correct, but would like a confirmation from @jtreuting.

File: common/src/main/java/org/opentestsystem/rdw/ingest/common/util/ParserHelper.java [hunk 9]
```diff
@@ -10,63 +13,117 @@
...
+     * @param rawValue raw field value
+     * @param parseFx  parsing function
+     * @param <T>      data type
+     * @return validated value of type T, may be null
      */
-    public ParserHelper(final DataElementErrorCollector elementErrorCollector) {
-        this.elementErrorCollector = elementErrorCollector;
+    public <T> T parseMandatory(final String name, final String rawValue, final Function<String, T> parseFx) {
+        try {
+            final T parsedValue = rawValue == null ? null : parseFx.apply(rawValue.trim());
```
Comment (malaffoon) [inline common/src/main/java/org/opentestsystem/rdw/ingest/common/util/ParserHelper.java]: Why the change from passing null to parseFx here? Before, the parseFx method could make decisions about null values, this takes that flexibility away.

File: exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultStudentExamProcessor.java [hunk 10]
```diff
@@ -115,41 +114,59 @@ public Exam parseExam(final TDSReport report, final int schoolId, final Assessme
...
+        Integer level = Integer.parseInt(value);
+        if (level > 0) return level;
+
+        throw new IllegalArgumentException("Invalid value, must be a positive number");
+    }
+
+    private int toValidYear(final Long value) {
+        if (value == null) throw new IllegalArgumentException("Invalid year");
+
+        if (value >= Year.MIN_VALUE && value <= Year.MAX_VALUE) return value.intValue();
```
Comment (agorina) [inline exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultStudentExamProcessor.java]: Do we want to be more restrictive here or is it good enough?

File: common/src/main/java/org/opentestsystem/rdw/ingest/common/util/ParserHelper.java [hunk 11]
```diff
@@ -10,63 +13,117 @@
...
+     * @param parseFx  parsing function
+     * @param <T>      data type
      * @return validated value of type T, may be null
      */
-    public <T> T validate(final String name, final String rawValue, final Function<String, T> fx) {
-        final T value = fx.apply(rawValue == null ? null : rawValue.trim());
-        if (value == null && elementErrorCollector != null)
-            elementErrorCollector.add(new DataElementError(name, rawValue, "Invalid value"));
-        return value;
+    public <T> T parseOptional(final String name, final String rawValue, final Function<String, T> parseFx) {
```
Comment (malaffoon) [inline common/src/main/java/org/opentestsystem/rdw/ingest/common/util/ParserHelper.java]: Hmm, i'll have to follow the code path. I'm not sure i like the changes here that take the optional/mandatory behavior away from the parse function.

File: exam-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/repository/impl/ExamRepositoryIT.java [hunk 12]
```diff
@@ -56,7 +56,7 @@
     @Autowired
     private JdbcTemplate jdbcTemplate;
 
-    //TODO: we do not have a referential integrity enforced on the db, hence we can use any data. We should consider adding system level reports to check for issues
```
Comment (agorina) [inline exam-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/repository/impl/ExamRepositoryIT.java]: I created a JIRA for this TODO.

File: common/src/main/java/org/opentestsystem/rdw/ingest/common/util/ParserHelper.java (hunk 9 above)
Comment (agorina) [inline common/src/main/java/org/opentestsystem/rdw/ingest/common/util/ParserHelper.java]: Not intentional, will add it back.

File: exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Exam.java [hunk 13]
```diff
@@ -128,7 +127,7 @@ public static Builder builder() {
         private long assessmentId;
         private String asmtVersion;
         private Long opportunity;
-        private Long schoolYear;
+        private Integer schoolYear;
```
Comment (malaffoon) [inline exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Exam.java]: Shouldn't the builder type be the same as the class type?

**Comment by agorina** [comment]
Talked with @malaffoon and will rework it based on the feedback.

File: common/src/main/java/org/opentestsystem/rdw/ingest/common/util/ParserHelper.java (hunk 9 above)
Comment (agorina) [inline common/src/main/java/org/opentestsystem/rdw/ingest/common/util/ParserHelper.java]: I think I am going to remove the functional programming approach.  Here is why:
```
  @Test
    public void test(){

        Function<String, String> fx = Precondition::checkNotBlank;
        Function<String, Integer> fx2 =  repository::findIdByCode;

        final Stopwatch stopwatch = Stopwatch.createStarted();

        int id = fx.andThen(fx2).apply("Male");

        System.out.println("execution time [" + stopwatch.toString() + "]");

        stopwatch.reset().start();

        id = repository.findIdByCode(Precondition.checkNotBlank("Male"));
        System.out.println("execution time [" + stopwatch.toString() + "]");
    }
```
execution time [6.082 ms]
execution time [666.9 μs]

Comment (malaffoon) [inline common/src/main/java/org/opentestsystem/rdw/ingest/common/util/ParserHelper.java]: No real objection here, i've been pretty consistently disappointed by java 7/8 functional performance. But did you try flipping the order; or alternate execution in a loop? It could very well be order of operation or ...?

Comment (malaffoon) [inline common/src/main/java/org/opentestsystem/rdw/ingest/common/util/ParserHelper.java]: So i did a similar example that doesn't involve the repo. Conclusion: functional may be slower but it might not be or may not be as bad as your test indicates.
{code}
    public void perfTest() {
        final Function<String, Integer> fx = Ints::tryParse;

        final Stopwatch stopwatch = Stopwatch.createUnstarted();

        long sum;
        final List<String> values = new Random().ints(1000).mapToObj(Integer::toString).collect(Collectors.toList());

        for (int t = 0; t < 3; ++t) {
            stopwatch.reset().start();
            sum = 0;
            for (final String value : values) {
                sum += Ints.tryParse(value);
            }
            stopwatch.stop();
            System.out.println("Direct, " + sum + ", " + stopwatch.toString());

            stopwatch.reset().start();
            sum = 0;
            for (final String value : values) {
                sum += fx.apply(value);
            }
            stopwatch.stop();
            System.out.println("Function, " + sum + ", " + stopwatch.toString());
        }
    }
{code}
I'm collecting the sum and printing it to try to convince the system to not optimize anything away. Results:
{code}
Direct, 34635744859, 2.275 ms
Function, 34635744859, 536.9 μs
Direct, 34635744859, 359.1 μs
Function, 34635744859, 390.2 μs
Direct, 34635744859, 508.3 μs
Function, 34635744859, 408.5 μs
{code}

Comment (agorina) [inline common/src/main/java/org/opentestsystem/rdw/ingest/common/util/ParserHelper.java]: You are right, it was just the order of the execution. Thank you. In this case I am going to keep it, unless you still think we should remove it?
